plugins/android_boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,410 @@
+		if (!fu_bytes_compare(blob1, blob2, error)) {
+			g_prefix_error(error,
+				       "failed to verify @0x%x: ",
+				       (guint)fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_android_boot_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* get data to write */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	if (g_getenv("FWUPD_ANDROID_BOOT_VERBOSE") != NULL)
+		fu_dump_bytes(G_LOG_DOMAIN, "write", fw);
+
+	if (g_bytes_get_size(fw) > fu_device_get_firmware_size_max(device)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "firmware too large, got 0x%x, expected <= 0x%x",
+			    (guint)g_bytes_get_size(fw),
+			    (guint)fu_device_get_firmware_size_max(device));
+		return FALSE;
+	}
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x0, 0x0, 10 * 1024);
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 72, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 20, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 7, NULL);
+
+	/* erase, write, verify */
+	if (!fu_android_boot_device_erase(self, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	if (!fu_android_boot_device_write(self, chunks, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	if (!fu_android_boot_device_verify(self, chunks, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	return TRUE;
+}
+
+static gboolean fu_android_boot_device_set_quirk_kv(FuDevice * device,
+						    const gchar *key,
+						    const gchar *value,
+						    GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+	guint64 tmp = 0;
+
+	/* load from quirks */
+	if (g_strcmp0(key, "AndroidBootVersionProperty") == 0) {
+		self->version = fu_android_boot_extract_cmdline_arg_value(value);
+		return TRUE;
+	}
+
+	if (g_strcmp0(key, "AndroidBootPartitionMaxSize") == 0) {
+		if (!fu_strtoull(value, &tmp, 0, G_MAXUINT32, error))
+			return FALSE;
+		self->max_size = tmp;
+		return TRUE;
+	}
+
+	/* failed */
+	g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED, "quirk key not supported");
+	return FALSE;
+}
+
+static void
+fu_android_boot_device_finalize(GObject *obj)
+{
+	G_OBJECT_CLASS(fu_android_boot_device_parent_class)->finalize(obj);
+}
+
+static void
+fu_android_boot_device_init(FuAndroidBootDevice *self)
+{
+	fu_device_set_summary(FU_DEVICE(self), "Android Bootloader Device");
```

Comment:
remove Device I think

Reply:
This is done in other plugins like this, it expects a FuDevice and not FuAndroidBootDevice.

Reply:
Ohh sorry for being unclear; i meant `Android Bootloader` should be the summary! :)